<html lang="en">
<head>
<meta charset="utf&#45;8">
<title>copurification.org</title>
<link rel="stylesheet" href="../copurification-html/main.css">
</head>
<body style="margin:10px;">

<h2>HOWTO</h2>

<p align="justify"><b>Overview &#45; How this works.</b></p>

<p align="justify">Anyone can view data that has been made public. Follow the prompts to retrieve results. Specific reagents can be included
as well as excluded from searches.
Multiple elements in search lists can be selected by Shift+click and Ctrl+click. We believe the functionality is self&#45;evident,
but should this not prove to be the case, please <script language="JavaScript">
	var username = "support"; var hostname = "copurification"; var linktext = "contact us";
	document.write("<a target='_blank' href='" + "mail" + "to:" + username + "@" + hostname + ".org" + "'>" + linktext + "</a>");
	</script>  &#45; we'll answer your query and add it to our FAQ if it proves to be common.

<p align="justify">In order to take advantage of the curation features of this site for your own work, you'll need to create an account. Instructions on how to complete each step and use this site follow below. The basics are these:

<p align="justify"><b>I.</b>	Create an account &#45; Ultimately this allows you to upload images of stained protein gels and files containing sample descriptions,
indicating which samples have been run on the gels. The images and sample descriptions are organized as "Experiments," which are associated with "Projects."
<br><br><b>II.</b>	Create a Project &#45; Projects are simply a way to organize your experiments. You must create a Project to add Experiments (images and sample descriptions) that will be associated with that Project. Multiple experiments can be added to any Project. Projects can also be created in any existing Project (i.e. as a sub&#45;Project). Create multiple Projects to keep your different kinds of affinity capture experiments organized.
<br><br><b>III.</b> Upload Experiments &#45; upload images of gels and tab&#45;delimited text files that list the samples loaded on the gels as well as
some basic descriptive text. Our software will section the gel image into individual lanes according to the number of lanes
specified in the text file. The experimental conditions used to derive that sample, also specified in the text file, will be
linked to the lane. In this way, different lanes from different gels can be compared, with the conditions used to obtain the
results also displayed. All your data are kept Private to your account until you choose to share them Publicly.
<br><br><b>IV.</b>	Retrieve results &#45; You can catalog all of your purifications, and track the resulting banding patterns. You can also view any results that others have
made Public and compare them to your own.  

<p align="justify"><i>As much as possible, we aim to implement <a target='_blank' href='http://mibbi.sourceforge.net/about.shtml'>Minimum Information Standards</a>  and
implement <a target='_blank' href='https://www.ebi.ac.uk/ontology&#45;lookup/init.do'>Ontologies and Controlled Vocabularies</a>  to maximize the value of experimental
data lodged in this repository.
</i>
	<br><br>

<p><b>HOWTO &#45; Create an account.</b><br><br>



<span style="text-align:justify;">

	Follow the link "Create Account" and fill out the web form. We encourage users to create an <a target='_blank' href='http://orcid.org'>ORCID</a> and
	associate this ID with their account.  These unique IDs provide a mechanism for a researcher to remain associated with and credited for their data and
	publications. Once logged in, your account details can be changed via the "My Account" link.</span><br><br>
	<img src="create_account.png"/ >
	
	

<br><br>
<p align="justify"><b>HOWTO &#45; Start a new Project.</b><br><br>

	While logged into your account, a new project may be added by clicking the "New Project" link. Once created, the new project will appear on the left side
	of the screen. Clicking on the project displays the "Create New Experiment" interface, where experimental details, including images of gels and the affinity
	capture conditions used, can be added to the project. A New Project can also be added to an existing Project by clicking on the Project and filling out the
	"Create New Project" interface.<br><br>
	<img src="create_project.png"/ >

<br><br>
<p align="justify"><b>HOWTO &#45; Add Experiments to a project.</b><br><br>

	Fill out the required fields, specifying the appropriate organism. "Experimental Procedures," a text file describing the
	experiments and protocols, may be uploaded but is not currently required (the system will accept TXT, RTF, and DOC files). A
	"Gel Details" file describing the conditions of the gel electrophoresis and basics of image development and capture
	may also be added; although currently not required, for this we strongly recommend you supply an
	<a target='_blank' href='http://www.psidev.info/protein&#45;separation'>MIAPE&#45;GE compliant file</a> (the system will accept TXT, RTF, DOC, and XML files).
	Experiments must include both a "Sample Descriptions file," listing the purification conditions used in the
	affinity capture experiments, and the associated gel image(s) showing the resulting banding patterns for each affinity captured sample. Currently,
	the Sample Descriptions file is a plain text file including tab&#45;delimited values. It must contain one line for each lane of the gel where a sample,
	marker, or quantity standard has been loaded. Each line contains numerical and categorical data (format described below). Upload as many images as needed,
	all the lanes present in each image are described in the same Sample Descriptions file. The images
	may be in TIFF, PNG, BMP, and JPEG format, but we recommend
	the use of 16&#45;bit TIFF images.  Please note that JPEG format images cannot be quantified by our band&#45;detection tool.</span><br><br>
	<img src="create_experiment.png"/ >
<br><br>
	Although our system will accommodate images in a variety of formats, resolutions, and bit&#45;depths, we recommend the use of unprocessed, uncompressed
	images; i.e. the original, raw data (e.g. as a 16&#45;bit TIFF). Our system will apply brightness and contrast calibration during the curation process without
	altering the original data  . To get the best results from our lane&#45;sectioning tool, gel images should be oriented such that the first lane (defined in
	the <a href='./Example-Samples_v09.txt' target='_blank'>Sample Descriptions</a> file) is at the left, and the last lane is  at the right, and that the gel is properly aligned
	<a href='./gel1.tif' target='_blank'>(see this example)</a>. For best results,
	image the gel in the appropriate orientation, initially. However, rotation and cropping of the gel image, when needed, can be carried out in an image&#45;processing
	program &#45; we recommend saving the file in the same format as the original.
<br><br>
	Our program is equipped to ignore empty lanes or otherwise omit lanes not described in the Sample Descriptions
	file. The total number of lanes in the image is defined in the Sample Descriptions file (example below). The positions of the samples loaded on the gel (lane number)
	are also defined in the same file.  Once the image is curated, the gel image is divided into lanes based on the total number specified in the Sample Descriptions
	file,
	and the lanes containing sample are identified based on their numerical position (counting from left to right), as specified in the Sample Descriptions file.
	The lane&#45;sectioning tool uses the average lane width as a guide to section empty lanes. Lanes not described in the Sample Descriptions file are treated as
	empty and ignored Do not include rows for empty lanes in the Sample descriptions file because certain descriptive columns are mandatory for all lanes in the file
	and will cause an error if not completed.  
<br><br>
<b>Example:</b> You've run four, twenty&#45;six&#45;well midi&#45;gels. Each gel contains one molecular weight marker, one BSA quantity standard and twenty&#45;four
experimental samples &#45; for a total of one hundred and four samples across four gels. You will need one <a href='./Nup1_example_data.txt' target='_blank'>
Sample Descriptions</a> file  listing all one hundred and four samples loaded across all four gels (see this example, and HOWTO below )
and four image files (one of each gel). Note: the names of the image files MUST match the "Gel file name" specified in the 
Sample Descriptions file  &#45; this is how our program knows which sample descriptions go with which gel and which lanes on that gel.
<br><br>
<p align="justify"><b>HOWTO &#45; Query Data in the System</b><br><br>

	  To query your data (public and private), use the 'Search My Gels' button  below the Projects listing in the left pane once logged in to the site.
	  Use the checkboxes to choose the Projects/Experiments to include in the Query.  <br><br>
	<img src="private_search.png"/ ><br><br>Public Data can be queried via the 'Search Public Gels' link in the
	  Navigation bar.  All query results will open in a new window so that your data may be compared to Public Data on the site.  <br><br>
	<img src="public_search.png"/ >

<br><br>
<p align="justify"><b><a id="SD">HOWTO</a> &#45; Fill the fields of the Sample Descriptions file.</b><br><br>

The <a href='./Sample_Descriptions_v09.txt' target='_blank'>Sample Descriptions</a> file  can be filled out as a spreadsheet in any program that can
save the result as a tab&#45;delimited text file,
such as MS Excel. The function of this file is to provide information about the samples loaded on the gel, such as the composition of the
extraction solvent used, the protein of interest (bait), the affinity tag (if any), the antibody used for capture, and so on. Many of the
short comings of the current system will vanish once we switch to an XML format &#45; and once we have had the chance to implement ontology and
controlled vocabulary. Please <script language="JavaScript">
	var username = "support"; var hostname = "copurification"; var linktext = "contact us";
	document.write("<a target='_blank' href='" + "mail" + "to:" + username + "@" + hostname + ".org" + "'>" + linktext + "</a>");
	</script>  for questions or doubts. We encourage adding references to certain fields as PMID, PMCID or
DOI when helpful, as discussed below.
<br><br>
When describing the extraction solvent(s) used, multiple reagents of the same class (most commonly salts and detergents) can be specified by
adding the appropriate headings multiple times (e.g. salt, salt concentration, salt unit for each different salt used). Note: we currently
accept the following units of metric concentration &#45; M, mM, uM, &mu;M, nM, pM. Note, micromolar units can use either a 'u' or a Greek letter mu as
the prefix. We also accept concentrations expressed as percent volume units: % w/v and % v/v (however, we are not currently processing
% w:v, % v:v or any units expressed as percent weight). <a href="./data_file_instructions.htm" target="_blank">Here</a>  is a full list of all the reagents
currently supported by our system. <script language="JavaScript">
	var username = "support"; var hostname = "copurification"; var linktext = "Contact us";
	document.write("<a target='_blank' href='" + "mail" + "to:" + username + "@" + hostname + ".org" + "'>" + linktext + "</a>");
	</script>
if you need to have a reagent added to this list for your upload to work. In most cases an example of common usage can be gleaned by viewing the public data and the
example
Sample Descriptions file.
<br>
<ul>
<li>The following fields are mandatory for each lane: <i>Gel file name</i>, <i>Lane # on gel</i>, and <i>Total # of lanes on gel</i>.</li>
<li>The following fields are mandatory for each lane EXCEPT those specifying a Mass Ladder or Quantity Standard:  <i>Over&#45;expressed</i>, <i>
Protein Systematic Name</i>, and <b>either</b> <i>Antibody</i> <b>or</b> <i>Other capture</i>.</li>
<li>Each gel must have at least one lane with an associated <i>Mass Ladder</i> column.     <i>Quantity Std</i> is optional for a gel, and
will be ignored if a JPEG file is submitted.</li>
<li>All fields permit 255 characters, except for notes, which permits 500 characters. </li>
</ul>

<p align="justify">These are the metadata contained within the file:<br><br>

<i>Gel file name</i>: specifies which image the entry applies to. The value entered here must match the file name of the gel image.<br><br>

<i>Lane # on gel</i>: specifies the position of the sample on the gel by lane number, from left to right.<br><br>

<i>Total # of lanes on gel</i>: specifies the total number of lanes.<br><br>

<i>Buffer</i>: specifies the agent primarily responsible for pH control of the extraction solvent.<br><br>

<i>Buffer Concentration</i>: the magnitude of the concentration.<br><br>

<i>Buffer Units</i>: the units of the concentration.<br><br>

<i>Salt</i>, <i>Detergent</i> and 'Other' additives follow the same convention as for <i>Buffer</i>. Other additives are commonly those that don't readily
classify as, or whose indented function in the solvent is distinct from, a buffering agent, salt or detergent (e.g. EDTA or urea).
<br><br>
<i>pH</i>: specifies the pH of the extraction solvent<br><br>

<i>Mass Ladder</i>: specifies the molecular masses, in kDa, of the bands present in the molecular weight marker; each mass is separated by a comma.<br><br>

<i>Quantity Std</i>: specifies the identity, molecular mass in kDa, quantity loaded, and quantity units of a standard protein; each entry is separated by a
comma. E.g. BSA, 66.5, 100, ng specifies that 100ng of BSA was loaded and that BSA is a protein of 66.5 kDa molecular mass.
<br><br>
<i>Over&#45;expressed</i>: specifies if the bait protein is over&#45;expressed in the sample or not. As a rule of thumb, we believe that an expression level
of more than double the wild&#45;type (wt) expression level could be considered over&#45;expressed. Amongst other things, knowing if the bait is over&#45;expressed
assists in assessing the significance of the levels of the other co&#45;purified proteins relative to the bait. Acceptable values for this field are yes, no, & unknown.
<br><br>
<i>Protein Systematic Name</i>: specifies the identity of the bait protein. We currently curate Saccharomyces cerevisiae,
Escherichia coli, and Homo sapiens genes and proteins. S. cerevisiae genes and proteins must be specified by Saccharomyces
Genome Database (SGD) systematic names. Human and E. coli genes and proteins must be specified by GenBank accession numbers;
we encourage the use of RefSeq accession numbers whenever appropriate for your bait sequence. In some cases the bait protein may
be present in multiple isoforms, or the isoform expressed may not be specifically defined &#45; such as when the expression construct
is not based on cDNA, but rather on e.g. gDNA containing intronic sequences and introducing the possibility of alternative splicing.
LAP&#45;tagged genes expressed via BAC TransgeneOmics (DOI: 10.1038/nmeth.1199) present such a case. The genomic locus of the protein of
interest is tagged and reintroduced, and the precise tagged protein expressed will depend on the cell type and may not be uniform.
In this or similar cases, where the precise protein product is less well defined, we encourage the use of GenBank Gene ID.
Additionally, in many cases the bait may be mutated or truncated by design. Where an SGD or GenBank record does not exist to
define the sequence utilized, use the sequence ID that most closely matches the expression construct and append additional info
in the Notes column to communicate the alterations. <script language="JavaScript">
	var username = "support"; var hostname = "copurification"; var linktext = "Contact us";
	document.write("<a target='_blank' href='" + "mail" + "to:" + username + "@" + hostname + ".org" + "'>" + linktext + "</a>");
	</script>  to add your organism to our database or for cases where you
encounter difficulties assigning an appropriate ID to your bait.
<br><br>
Helpful links:<br>
<a target='_blank' href='http://www.yeastgenome.org'>http://www.yeastgenome.org</a> <br>
<a target='_blank' href='http://www.ncbi.nlm.nih.gov/protein'>http://www.ncbi.nlm.nih.gov/protein</a> <br>
<a target='_blank' href='http://www.ncbi.nlm.nih.gov/gene'>http://www.ncbi.nlm.nih.gov/gene</a><br>
<a target='_blank' href='http://www.ncbi.nlm.nih.gov/refseq'>http://www.ncbi.nlm.nih.gov/refseq</a> <br><br>

<i>Tag Type</i>: if the bait is captured through an affinity tag, this entry provides for the tag to be specified. Please enter the name of the tag and any additional info, especially references (as PMID, PMCID or DOI), in parentheses. Some common tags include: LAP, GFP, FLAG, 3xFLAG, Myc, SpA (based on wt Protein A), TAP, ZZ (tandem synthetic Protein&#45;A&#45;based Z&#45;domains), 3xHIS (and various other multiple), etc. and so on.
<br><br>
<i>Tag Location</i>: if the bait is tagged, this specifies where. Acceptable answers are N&#45;term, C&#45;term, and Internal. Required if Tag Type is specified.
<br><br>
<i>Antibody</i>: specifies the antibody reagent used to capture the bait protein, either directly or through an appended affinity tag. Provide a short description of the reagent &#45; such as Rabbit IgG (which might be use for purifying a Protein&#45;A&#45;tagged protein) or polyclonal llama anti&#45;GFP, or anti&#45;FLAG M2. In parentheses please provide a manufacturer and product number if purchased commercially, and or an appropriate reference for the reagent. 
<br><br>
<i>Other Capture</i>: if the affinity reagent used was not an antibody, or if multiple affinity steps were utilized (e.g. including a second reagent, antibody or not, such as in tandem affinity purification), then the reagent used may be specified here. Follow the conventions used above.
<br><br>
<i>Notes</i>: in this section any additional details important for understanding the affinity capture may be provided. Please be judicious and succinct. Examples of info to provide might include details of the construct or expression system, capture method, etc. References are always encouraged.  
<br><br>
<p align="justify"><b><a id="SD">HOWTO</a> &#45; Lane Clustering.</b><br><br>

When logged in to the system and viewing an Experiment, click on the link "View Lane Clustering."  This will open a new page in the web browser with the lanes for each captured protein clustered by matching bands.  Hovering the mouse pointer over each image will open a tooltip listing the Reagents for a lane.

<br><br>
	<img src="lane_clustering.png"/ ><br><br>
</body>
</html>